d2/src/main/java/com/linkedin/d2/balancer/util/HostOverrideList.java

```diff
@@ -0,0 +1,103 @@
+/*
+   Copyright (c) 2020 LinkedIn Corp.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package com.linkedin.d2.balancer.util;
+
+import com.linkedin.util.ArgumentUtil;
+import java.net.URI;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
+
+
+/**
+ * Stores the list of host overrides for cluster and services. The order of the overrides are stored in the same
+ * order as the additions. Checks for the first match and return the overridden {@link URI}.
+ */
+public class HostOverrideList
+{
+  private Map<Key, URI> _overrides = new LinkedHashMap<>();
+
+  public void addClusterOverride(String cluster, URI uri) {
+    ArgumentUtil.notNull(cluster, "cluster");
+    ArgumentUtil.notNull(uri, "uri");
+    _overrides.put(new Key(cluster, null), uri);
```

Comment:
Since the overrides could be supplied by different app, what if there are conflicts?

Reply:
It won't be supplied by different apps since only the current app can add the override. However, it may be supplied by different places in the code. A solution I can think of is having a state of the list such that the first caller can add to the list and seal the list. The second caller will receive an exception adding to a sealed list. Another option is to expose the list contents to the caller so they can decide where to add the override. I don't particularly like either one as both add significant complexity to the logic.